llmfoundry/data/finetuning/tasks.py

```diff
@@ -70,9 +98,70 @@ def _is_empty_or_nonexistent(dirpath: str) -> bool:
     return not os.path.isdir(dirpath) or len(os.listdir(dirpath)) == 0
 
 
-def _tokenize_formatted_example(
-        example: Dict[str, Any],
-        tokenizer: PreTrainedTokenizerBase) -> Dict[str, List[int]]:
+def _slice_chat_formatted_example(
+        example: ChatFormattedDict,
+        tokenizer: PreTrainedTokenizerBase) -> Tuple[str, str]:
+    """
+    Applies the tokenizer's chat template to the example messages and slices the resulting templated string into a prompt and a completion.
+    
+    Args:
+        example (ChatFormattedDict): The chat example containing the messages.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer to apply the chat template.
+    
+    Returns:
+        Tuple[str, str]: The prompt and response as separate strings.
+    
+    Raises:
+        ValueError: If the chat example has less than two messages or if the last message is not from the assistant.
+        KeyError: If a message does not have a role or content.
+    """
+    def slice(s: str, sep: str):
+        # it seems like we can reuse this logic, as we likely have this pattern in other places.
+        slices = s.split(sep)
+        if len(slices) < 2:
+            raise ValueError(f'separator not in string. {sep=}, {s=}')
+        a, b = sep.join(slices[:-1]), sep + slices[-1]
+        return a, b
+
+    messages = example['messages']
+
+    if len(messages) < 2:
+        raise ValueError(
+            f'chat example must have at least two messages. {messages=}')
+    last_message = messages[-1]
+    if last_message['role'] != 'assistant':
+        raise ValueError(
+            f'last message must be from assistant. {last_message=}')
+    for message in messages:
+        if 'role' not in message or 'content' not in message:
+            raise KeyError(f'message must have role and content. {message=}')
+
+    applied_template = tokenizer.apply_chat_template(messages, tokenize=False)
+    prompt, response = slice(applied_template, last_message['content'])
+    return prompt, response
```

Comment:
maybe appropriate to add some checks here? like response not empty or something

just in case there are tokenizer issues like: https://huggingface.co/mosaicml/mpt-7b-8k-chat/discussions/8 haha